### PR TITLE
[refactor] Update Ln-norm logic for upcoming PyTorch update

### PR DESCRIPTION
**Pull request by namgyu-youn**
## What does this PR do?

**Type of change:** Refactor 

**Overview:** Based on [PyTorch - docs](https://docs.pytorch.org/docs/stable/generated/torch.norm.html), `torch.norm()` will be removed in the foreseeable version. Therefore, this update replaces `torch.norm()` with `torch.linalg.vector_norm()`

## Usage

```python
# Old (current) version
torch.norm()

# New (upcoming) version
torch.linalg.vector_norm()
```

## Testing

pytest and the comment

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)**
- **Is this change backward compatible?**: Yes  
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: No

### CODE REVIEW

**Comment by namgyu-youn** [comment]
cc. @kevalmorabia97 @meenchen

**Comment by kevalmorabia97** [comment]
Thanks @namgyu-youn for contributing the fix. Can you also apply similar changes to usage of `<tensor>.norm()` throughout the repository?

Please also share testing details - whether you have run unit and GPU tests? (`pytest tests/unit` and `pytest tests/gpu`)

**Comment by namgyu-youn** [comment]
> Thanks @namgyu-youn for contributing the fix. Can you also apply similar changes to usage of `<tensor>.norm()` throughout the repository?

Sure, and it seems there is no other Ln-norm logic in `modelopt/`. Let me know if I am missing something.

> Please also share testing details - whether you have run unit and GPU tests? (`pytest tests/unit` and `pytest tests/gpu`)

The result of pytest is the following:
image

Also, local test code is the following:
```python
import torch

# Create test tensors
torch.manual_seed(42)
X = torch.randn(10, 10)
X_hat, A_reg = torch.randn(10, 10), torch.randn(5, 5)

# Old Version (torch.norm)
print("=== Old Version (torch.norm) ===")
old_relative_error = torch.dist(X, X_hat, p=torch.inf) / torch.norm(X, p=torch.inf)
old_norm_result = torch.norm(A_reg, p=torch.inf)

print("Relative error:", old_relative_error.item())
print("A_reg infinity norm:", old_norm_result.item())
print()

# New Version (torch.linalg.vector_norm)
print("=== New Version (torch.linalg.vector_norm) ===")
new_relative_error = torch.dist(X, X_hat, p=torch.inf) / torch.linalg.vector_norm(X, ord=torch.inf)
new_norm_result = torch.linalg.vector_norm(A_reg, ord=torch.inf)

print("Relative error:", new_relative_error.item())
print("A_reg infinity norm:", new_norm_result.item())
```
And the result is the following:
```bash
=== Old Version (torch.norm) ===
Relative error: 1.512144684791565
A_reg infinity norm: 1.9775909185409546

=== New Version (torch.linalg.vector_norm) ===
Relative error: 1.512144684791565
A_reg infinity norm: 1.9775909185409546
```

**Comment by namgyu-youn** [comment]
> Thanks @namgyu-youn for contributing the fix. Can you also apply similar changes to usage of `<tensor>.norm()` throughout the repository?

@kevalmorabia97; Sure, and it seems there is no other Ln-norm logic in `modelopt/`. Let me know if I am missing something.

> Please also share testing details - whether you have run unit and GPU tests? (`pytest tests/unit` and `pytest tests/gpu`)

The result of pytest is the following:
image

Also, local test code is the following:
```python
import torch

# Create test tensors
torch.manual_seed(42)
X = torch.randn(10, 10)
X_hat, A_reg = torch.randn(10, 10), torch.randn(5, 5)

# Old Version (torch.norm)
print("=== torch.norm ===")
relative_error = torch.dist(X, X_hat, p=torch.inf) / torch.norm(X, p=torch.inf)
norm_result = torch.norm(A_reg, p=torch.inf)

print("Relative error:", relative_error.item())
print("A_reg infinity norm:", norm_result.item())
print()

# New Version (torch.linalg.vector_norm)
print("=== torch.linalg.vector_norm ===")
relative_error = torch.dist(X, X_hat, p=torch.inf) / torch.linalg.vector_norm(X, ord=torch.inf)
norm_result = torch.linalg.vector_norm(A_reg, ord=torch.inf)

print("Relative error:", relative_error.item())
print("A_reg infinity norm:", norm_result.item())
```
And the result is:
```bash
=== torch.norm ===
Relative error: 1.512144684791565
A_reg infinity norm: 1.9775909185409546

=== torch.linalg.vector_norm ===
Relative error: 1.512144684791565
A_reg infinity norm: 1.9775909185409546
```

**Comment by kevalmorabia97** [comment]
Please see https://github.com/search?q=repo%3ANVIDIA%2FTensorRT-Model-Optimizer+.norm%28&type=code

**Comment by namgyu-youn** [comment]
> Please see https://github.com/search?q=repo%3ANVIDIA%2FTensorRT-Model-Optimizer+.norm%28&type=code

Got it, really thanks for your leading.

**Comment by kevalmorabia97** [comment]
Thanks for the changes. Let me take this in our internal repo where we run a more extensive CI test which is yet to be migrated to Github.

File: modelopt/torch/nas/plugins/megatron.py [hunk 1]
```diff
@@ -622,11 +622,15 @@ def _estimate_all_head_importance(self) -> TracedHp.Importance:
     def _estimate_query_group_importance(self) -> TracedHp.Importance:
         """Return the importance of the ``num_query_groups`` hparam."""
         assert self._activations is not None, "No activations collected for importance estimation."
-        group_importance = self._activations.view(
-            self.get_hparam("num_heads_per_group").max,
-            self.get_hparam("num_query_groups").max,
-            self.config.kv_channels,
-        ).norm(p=2, dim=(0, 2))
+        group_importance = torch.linalg.norm(
```
Comment (kevalmorabia97) [inline modelopt/torch/nas/plugins/megatron.py]: Can you check if we need `torch.linalg.vector_norm` here? I tested a 3d random tensor and `torch.linalg.norm` seems to be different than previous `torch.norm`

File: modelopt/torch/nas/plugins/transformers.py [hunk 2]
```diff
@@ -122,7 +122,9 @@ def configure_qkv_out(self, q_name: str, k_name: str, v_name: str, out_name: str
         out.in_features = hp_hidden_dim
 
         assert isinstance(out, nn.Linear)
-        hp_hidden_dim.register_importance(lambda: out._parameters["weight"].detach().norm(dim=0))
+        hp_hidden_dim.register_importance(
+            lambda: torch.linalg.norm(out._parameters["weight"].detach(), dim=0)
```
Comment (kevalmorabia97) [inline modelopt/torch/nas/plugins/transformers.py]: Should we use vector_norm here as well?

File: modelopt/torch/nas/plugins/megatron.py (hunk 1 above)
Comment (namgyu-youn) [reply modelopt/torch/nas/plugins/megatron.py]: In [PyTorch docs](https://docs.pytorch.org/docs/stable/generated/torch.linalg.norm.html#torch.linalg.norm), `torch.linalg.norm()` supports various calculations based on `dim`:

```
- If dim is an int, the vector norm will be computed.
- If dim is a 2-tuple, the matrix norm will be computed.
- If dim=None and ord=None, A will be flattened to 1D and the 2-norm of the resulting vector will be computed.
- If dim=None and ord!=None, A must be 1D or 2D.
```

Even though we need a **vector norm** here, `dim=(0,2)` forces it to compute a matrix norm. Therefore, `torch.linalg.vector_norm()` is the correct choice, as you mentioned.

File: modelopt/torch/nas/plugins/transformers.py (hunk 2 above)
Comment (namgyu-youn) [reply modelopt/torch/nas/plugins/transformers.py]: This case computes vector norm because `dim` is int. Please check [above comment](https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/206#discussion_r2156059314); let me know if there is anything wrong in internal CI

**Comment by kevalmorabia97** [comment]
There is one more torch.norm change missing at https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/bb630dbc29197e99bf3e719f7101c2676aa520c8/modelopt/torch/nas/plugins/megatron.py#L616-L619
Can you please address that as well?

File: modelopt/torch/nas/plugins/transformers.py (hunk 2 above)
Comment (kevalmorabia97) [reply modelopt/torch/nas/plugins/transformers.py]: Lets use vector_norm explicitly to avoid any ambiguity and consistency with rest of the changes

**Comment by namgyu-youn** [comment]
> There is one more torch.norm change missing at
> 
> https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/bb630dbc29197e99bf3e719f7101c2676aa520c8/modelopt/torch/nas/plugins/megatron.py#L616-L619
> 
> 
> Can you please address that as well?

@kevalmorabia97 ; Thanks for your leading. Could you take a look at this PR?

**Comment by namgyu-youn** [comment]
Also, could you comment about [#146 - Comment](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/146#issuecomment-2996256492)? Reviewing source codes in `modelopt/torch/nas` and `modelopt/torch/prune` was the start for a deeper (new feature) contribution.

But before the start, your (and other maintainers) advice must be valuable background for the directionality. I hope the ticket is still open.

**Comment by kevalmorabia97** [comment]
Your PR commit is merged. Thank you for contributing
